torch/_jit_internal.py

```diff
@@ -823,33 +823,28 @@ def is_dict(ann) -> bool:
         (getattr(ann, '__origin__', None) is Dict or
             getattr(ann, '__origin__', None) is dict)
 
-def is_optional(ann) -> bool:
-    if ann is Optional:
-        raise_error_container_parameter_missing("Optional")
+def is_union(ann):
+    if ann is Union:
+        raise_error_container_parameter_missing("Union")
 
-    # Optional[T] is just shorthand for Union[T, None], so check for both
-    def safe_is_subclass(the_type, super_type):
-        # Don't throw if `the_type` isn't a class type (e.g. if it is
-        # another type annotation instance)
-        if not inspect.isclass(the_type):
-            return False
-        return issubclass(the_type, super_type)
+    return (hasattr(ann, '__module__') and
+            ann.__module__ == 'typing' and
+            (getattr(ann, '__origin__', None) is Union))
 
-    if not hasattr(ann, '__module__'):
-        return False
+def is_optional(ann):
+    if ann is Optional:
+        raise_error_container_parameter_missing("Optional")
 
-    union_optional = False
```

Comment:
I've looked at the code as well as the blame on Phabricator, and I'm not sure how this could break BC. To me, it seems as though it's a way to see if we have `Union[T, None`, which can be parsed as `Optional`. Could you please let me know which Python version this was put in here for?

Reply:
afaik `Union[T, None]` and `Optional[T]` have always been equivalent to Python, but at some point they switched from desugaring to `Union` to `Optional`:

python 3.6

```python
>>> import typing
>>> print(typing.Union[int, None])
typing.Union[int, NoneType]
>>> print(typing.Optional[int])
typing.Union[int, NoneType]
>>> print(typing.Union[int, None] == typing.Optional[int])
True
```

python 3.8

```python
>>> import typing
>>> print(typing.Union[int, None])
typing.Optional[int]
>>> print(typing.Optional[int])
typing.Optional[int]
>>> print(typing.Union[int, None] == typing.Optional[int])
True
```

Reply:
Now that we've introduced Union, we should be fine whether we have 3.6 or 3.8, though